### PR TITLE
Langchain destination: Add Chroma support

### LINKED ISSUE #28516: Langchain destination: Add Chroma support

### PR DESCRIPTION
**Pull request by flash1293**
## What

Fixes https://github.com/airbytehq/airbyte/issues/28516

## How

Adds a new indexing option for [Chroma](https://www.trychroma.com/). This is similar to the docarray option as it's saving the vector db in a local file.

I had to switch the base image to python 3.10, because the sqlite version would be too old for chroma to work properly otherwise.

### CODE REVIEW

**Comment by octavia-squidington-iii** [comment]
##  destination-langchain test report (commit [`e672c087fb`](https://github.com/airbytehq/airbyte/pull/28605/commits/e672c087fbdc3528c724c8ea54645b986bee3b93)) - ✅

⏲️  Total pipeline duration: 04mn12s 

| Step                                                                         | Result   |
|:-----------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-langchain/metadata.yaml | ✅        |
| Connector version semver check                                               | ✅        |
| QA checks                                                                    | ✅        |
| Code format checks                                                           | ✅        |
| Connector package install                                                    | ✅        |
| Build destination-langchain docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                   | ✅        |
| Integration tests                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_langchain-chroma/1690197811/e672c087fbdc3528c724c8ea54645b986bee3b93/destination-langchain/0.0.5/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-langchain test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-langchain test report (commit [`e672c087fb`](https://github.com/airbytehq/airbyte/pull/28605/commits/e672c087fbdc3528c724c8ea54645b986bee3b93)) - ✅

⏲️  Total pipeline duration: 12.51s 

| Step                                                                         | Result   |
|:-----------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-langchain/metadata.yaml | ✅        |
| Connector version semver check                                               | ✅        |
| QA checks                                                                    | ✅        |
| Code format checks                                                           | ✅        |
| Connector package install                                                    | ✅        |
| Build destination-langchain docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                   | ✅        |
| Integration tests                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_langchain-chroma/1690198143/e672c087fbdc3528c724c8ea54645b986bee3b93/destination-langchain/0.0.5/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-langchain test
```

File: airbyte-integrations/connectors/destination-langchain/setup.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
     "typing_extensions==4.5.0",
     "pydantic==1.10.8",
     "pandas==1.4.2",
+    "chromadb",
```
Comment (girarda) [inline airbyte-integrations/connectors/destination-langchain/setup.py:20]: should we pin the version?

File: airbyte-integrations/connectors/destination-langchain/destination_langchain/indexer.py [hunk 2]
```diff
@@ -140,3 +141,37 @@ def check(self) -> Optional[str]:
...
+    def pre_sync(self, catalog: ConfiguredAirbyteCatalog):
+        self._init_vectorstore()
+        for stream in catalog.streams:
+            if stream.destination_sync_mode == DestinationSyncMode.overwrite:
+                self.vectorstore._collection.delete(where={METADATA_STREAM_FIELD: {"$eq": stream.stream.name}})
+
+    def index(self, document_chunks, delete_ids):
+        for delete_in in delete_ids:
+            self.vectorstore._collection.delete(where={METADATA_RECORD_ID_FIELD: {"$eq": delete_in}})
+        self.vectorstore.add_documents(document_chunks)
```
Comment (girarda) [inline airbyte-integrations/connectors/destination-langchain/destination_langchain/indexer.py:168]: I might have misconfigured my connection, but my test sync failed on this line

```
  File "/airbyte/integration_code/destination_langchain/destination.py", line 65, in write
    batcher.flush()
  File "/airbyte/integration_code/destination_langchain/batcher.py", line 21, in flush
    self.flush_handler(list(self.buffer))
  File "/airbyte/integration_code/destination_langchain/destination.py", line 59, in <lambda></lambda>
    batcher = Batcher(BATCH_SIZE, lambda batch: self._process_batch(batch))
  File "/airbyte/integration_code/destination_langchain/destination.py", line 51, in _process_batch
    self.indexer.index(documents, ids_to_delete)
  File "/airbyte/integration_code/destination_langchain/indexer.py", line 168, in index
    self.vectorstore.add_documents(document_chunks)
  File "/usr/local/lib/python3.10/site-packages/langchain/vectorstores/base.py", line 104, in add_documents
    return self.add_texts(texts, metadatas, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/langchain/vectorstores/chroma.py", line 193, in add_texts
    self._collection.upsert(
  File "/usr/local/lib/python3.10/site-packages/chromadb/api/models/Collection.py", line 298, in upsert
    ids, embeddings, metadatas, documents = self._validate_embedding_set(
  File "/usr/local/lib/python3.10/site-packages/chromadb/api/models/Collection.py", line 357, in _validate_embedding_set
    validate_metadatas(maybe_cast_one_to_many(metadatas))
  File "/usr/local/lib/python3.10/site-packages/chromadb/api/types.py", line 169, in validate_metadatas
    validate_metadata(metadata)
  File "/usr/local/lib/python3.10/site-packages/chromadb/api/types.py", line 140, in validate_metadata
    raise ValueError(
ValueError: Expected metadata value to be a str, int, or float, got False which is a <class></class>
```

I used the Stripe source. Here is my destination config:
Screen Shot 2023-07-25 at 11 43 05 AM

**Comment by flash1293** [comment]
@girarda Thanks for catching this, it looks like Chroma doesn't like boolean values as meta data, added some logic to convert them into strings

**Comment by octavia-squidington-iii** [comment]
##  destination-langchain test report (commit [`fede6610b7`](https://github.com/airbytehq/airbyte/pull/28605/commits/fede6610b7075cc820d23f1cd9d142bc4f7573c7)) - ✅

⏲️  Total pipeline duration: 07mn02s 

| Step                                                                         | Result   |
|:-----------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-langchain/metadata.yaml | ✅        |
| Connector version semver check                                               | ✅        |
| QA checks                                                                    | ✅        |
| Code format checks                                                           | ✅        |
| Connector package install                                                    | ✅        |
| Build destination-langchain docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                   | ✅        |
| Integration tests                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_langchain-chroma/1690369307/fede6610b7075cc820d23f1cd9d142bc4f7573c7/destination-langchain/0.0.5/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-langchain test
```

**Review by girarda — APPROVED** [review]
looks good and I was able to write to Chroma :shipit:

**Pull request merged**